util-app/src/main/scala/com/twitter/app/App.scala

```diff
@@ -373,8 +383,13 @@ trait App extends Closable with CloseAwaitably {
     close(defaultCloseGracePeriod)
 
     // The deadline to 'close' is advisory; we enforce it here.
-    if (!suppressGracefulShutdownErrors) Await.result(this, closeDeadline - Time.now)
-    else {
+    if (!suppressGracefulShutdownErrors) {
```

Comment:
If seems we could replace all of this with:

```
    // The deadline to 'close' is advisory; we enforce it here.
    try {
      Await.result(this, closeDeadline - Time.now)
    } catch {
      case NonFatal(exc) =>
        if (!suppressGracefulShutdownErrors) {
          exitOnError(newCloseException(Seq(exc)))
        } else Unit
    }
```
I think the issue continues to be a way to test this.

Reply:
The point of the `suppressGracefulShutdownErrors` was to make sure the process exited with a successful exit code (0) and that any error here did not bubble up.

Reply:
I understand the idea of `suppressGracefulShutdownErrors`, however it won't call `System.exit(0)` and only rely on the `main` finishing. So if there is a non-daemon thread still running because of a closing exception, the app will never be closed. That's what I meant [here](https://github.com/twitter/util/pull/251#discussion_r308465982).

You say that we need a way to test this. Is there anything wrong with the jittered timer approach? If so what are you suggesting?

Reply:
The current machinery is meant to throw an exception due to the result of `Await.result` or to try to let any draining happen with the `Await.ready` (which is just checking if the `this` Future is eventually satisfied). The semantics of the two are subtly different.

In the first case, it is expected that any exception bubbles up the stack and is caught and handled by anything that cares to perform a clean shutdown -- otherwise it uncaught and the JVM exits. Whereas the latter case is purposely intended to not bubble up any errors on closing in this manner as it is expected that the orchestration layer is tearing things down anyway and, yes, relies on the main simply exiting.

There's a lot of machinery built on this and how this works, so changing things here is generally not simple. Even my example above fails for lots of reasons under testing.

It sounds like you're in a situation where you have started daemon threads and an exception happens when trying to close them which prevents the parent process from shutting down. I don't think there is much we can do here in the App code to help in this situation. Trying to signal that the app should terminate the JVM is unfortunately not possible in a non API breaking way currently.

My suggestion would be to create a wrapper process which calls to `App.main` in a try/catch which then can call System.exit for you.

Reply:
Another suggestion is to introduce another API to signal that you explicitly want to "exit" on error. We use `App` primarily through `TwitterServer` internally and in the semantics of that usage are to call `close()` manually on the server as it is blocking in the call to main in `App` [here](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L363). We expect the main to then exit accordingly without terminating the JVM with a `System.exit` call.

We could introduce a boolean, say, `def exitOnError: Boolean = false` which could be overridden to `true` in your application. It would trigger the `exitOnError(throwable)` when an exception is encountered and application has signaled this should exit the JVM.

Note: I'm sure we can re-use `exitOnError` like this but it's as good a name for the variable as  any.

Reply:
I have to admit, I am also still confused about the use case you are trying to solve. The call to `nonExitingMain` is done by `main` and is wrapped in a try/catch [here](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L331-L340). In the patched code, you are handling a `TimeoutException` which should be emitted from the `nonExitingMain` and intercepted by the case in the try/catch in main. Which would then terminate through one of two code paths: either directly [here](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L71-L74) due to getting a `CloseException` or by trying to wait a little longer for the future to complete and then bailing [here](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L91-L97).

Based on your patch, I'm not sure why you'd need to touch the code in the `nonExitingMain` which is maybe why I'm confused about the situation you're trying to handle. I could see trying to patch the `Await.ready` [here](https://github.com/twitter/util/blob/develop/util-app/src/main/scala/com/twitter/app/App.scala#L96) if things are hung to make sure it continues to the `System.exit`. But I would think that if you're trying to get the process to exit due to a `TimeoutException` on either the `Await.result` or `Await.ready` calls in the `nonExitingMain`, it does?

Also, I meant to answer you about the "jittered timer approach" -- yes, my concerns would be that test would be flakey in the sense it would not be deterministic.